### PR TITLE
Restws 895: Ensure fetching Allergies for a newly created patient returns HTTP 204-No Content Response Status code

### PR DESCRIPTION
**Pull request by mherman22**
## Description of what I changed

The main objective of this pull request is to ensure that while fetching allergies for a patient with no allergies, the response body should be empty and the status code should be [HTTP 204 NO CONTENT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) [as per the discussion](https://talk.openmrs.org/t/newly-registered-patients-default-allergy-value/37228/19).
- I added a new class called `NoContentFoundException` that extends the `ResponseException` class.
- I added the @ResponseStatus annotation whose value i set to HttpStatus' `HTTP 204 No Content`.
- I then ensured that in the `doGetAll` method in the `PatientAllergyResource2_0` class throws the  `NoContentFoundException` whenever Allergies is `unknown`

## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-895

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by dkayiwa** [comment]
@mherman22 did you notice that these changes contradict the pull request title?

**Comment by dkayiwa** [comment]
@mherman22 how is this test ensuring that http status 204 is returned?

**Comment by mherman22** [comment]
> @mherman22 how is this test ensuring that http status 204 is returned? Even if i reverted all your changes in `PatientAllergyResource2_0.java`, your test would still pass. Confirming that you are not testing your changes.

thanks @dkayiwa , kindly take a look again!

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java [hunk 1]
```diff
@@ -244,6 +247,18 @@ public void shouldThrowExceptionWhenSettingNoKnownAllergiesIfAllergiesNotEmpty()
...
+	/**
+	 * Return no response body with a 204 http response status code for a patient with UNKNOWN Allergies
+	 */
+	@Test(expected = NoContentFoundException.class)
+	public void shouldEnsureFetchingAllergiesForPatientReturnsNoContentResponseStatusCode() throws Exception {
+		handle(newDeleteRequest(getURI(), new Parameter("reason", "unit test")));
+		
+		MockHttpServletResponse response = handle(newGetRequest(getURI()));
+		
+		Assert.assertEquals(HttpStatus.NO_CONTENT.value(), response.getStatus());
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: If the`NoContentFoundException` is thrown, will the above line ever be reached?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: i believe when the [NoContentFoundException](https://github.com/openmrs/openmrs-module-webservices.rest/blob/d1ac1e1f868249f87cc613ef3220bb63882682f9/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/response/NoContentFoundException.java) is thrown [here](https://github.com/openmrs/openmrs-module-webservices.rest/blob/d1ac1e1f868249f87cc613ef3220bb63882682f9/omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/PatientAllergyResource2_0.java#L172), its an indication that the response status code is a [204](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) resulting from the patient lacking an allergy.

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: Did you try run the test in debug mode and confirmed that line 260 is hit?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: lemi try that, thanks!

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: > If the`NoContentFoundException` is thrown, will the above line ever be reached?

when i remove `@Test(expected = NoContentFoundException.class)` from the test, the test fails when `NoContentFoundException` is thrown.

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: You still have not answered the question.

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: to answer the question, the mentioned line (260) wouldnt be reached. on going through the reporters talk post, i need to return a `204 No Content with empty body` which i have only achieved halfway. does that answer the question @dkayiwa ?

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: Is the `Assert.assertEquals` on line 260 ever executed?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: yes it is

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: Can you run the test in debug mode and confirm that you can step over this line?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: i ran the test in debug mode as earlier suggested, i added a toggle breakingpoint but nothing happens. everything is blank. no failure trace given nothing. 

this goes to all other tests tht i tried to do the same.

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: The other alternative would be to temporarily put `System.out.println()` on various lines and see the out put on the console to see which lines were run and which were not. With that, you do not have to run in debug mode.

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: i have tried this approach but for some reason i dont get any results in the console. I could be doing something wrong

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: If you removed the throwing of the exception on line 172 of PatientAllergyResource2_0.java, are you then able to get the results on the console?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: when i remove the throwing on line 172, i get this error in the console which says `Caused by: java.lang.AssertionError: expected:<204> but was:<200>`.






The whole failure trace is shown below:
```
java.lang.Exception: Unexpected exception, expected<org.openmrs.module.webservices.rest.web.response.NoContentFoundException> but was<java.lang.AssertionError>
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:28)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:73)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:82)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:73)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:217)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:83)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:68)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:163)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.lang.AssertionError: expected:<204> but was:<200>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at org.openmrs.module.webservices.rest.web.v1_0.controller.openmrs2_0.PatientAllergyController2_0Test.shouldEnsureFetchingAllergiesForPatientReturnsNoContentResponseStatusCode(PatientAllergyController2_0Test.java:260)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	... 24 more


```

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: Doesn't that confirm that the line is not executed with your changes to fix the reported problem on the ticket?

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: Test method expects a `204` response which is provided for in the `nocontentfoundexception()` method which results into the test being green. However when i remove the exception from [line 172](https://github.com/openmrs/openmrs-module-webservices.rest/blob/d1ac1e1f868249f87cc613ef3220bb63882682f9/omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/PatientAllergyResource2_0.java#L172), i get a `200` reponse which results into the test failing. My thinking is, the line 260 in the test method is hit

Comment (dkayiwa) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java:260]: > My thinking is, the line 260 in the test method is hit

Basing on how your test is written, line 260 is not hit unless you can prove so.

Comment (mherman22) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/PatientAllergyController2_0Test.java]: > test is written

look through again please

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/response/NoContentFoundException.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.webservices.rest.web.response;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+/**
+ * This exception should be thrown from controllers when passed a patient uuid of UNKNOWN allergy.
```
Comment (dkayiwa) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/response/NoContentFoundException.java:16]: Is this exception to be used for only patients and allergies?

Comment (mherman22) [reply omod-common/src/main/java/org/openmrs/module/webservices/rest/web/response/NoContentFoundException.java]: @dkayiwa i've changed the wording in the latest commit

**Pull request merged**